File: src/main/java/offer/Offer26.java
package offer;

/**
 * @author jingxinwu
 * @date 2021-11-21 6:22 下午
 */
public class Offer26 {


    //子树
    public boolean isSubStructure(TreeNode A, TreeNode B) {
        if (A == null && B == null) {
            return true;
        }

        if (A == null) {
            return false;
        }
        if (B == null) {
            return true;
        }

        if (A.val == B.val) {
            return true;
        }

        return isSubStructure(A.left, B.left) || isSubStructure(A.right, B.right);
    }


    public boolean isSubStructureV2(TreeNode A, TreeNode B) {
        return (A != null && B != null) && (recur(A, B) || isSubStructure(A.left, B) || isSubStructure(A.right, B));
    }

    boolean recur(TreeNode A, TreeNode B) {
        if (B == null) {
            return true;
        }
        if (A == null || A.val != B.val) {
            return false;
        }
        return recur(A.left, B.left) && recur(A.right, B.right);
    }
}

File: src/main/go/leetcode/lists/offer/Offer26.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/3 18:52
 */

func isSubStructure(A *TreeNode, B *TreeNode) bool {

	var dfs func(A *TreeNode, B *TreeNode) bool
	dfs = func(A *TreeNode, B *TreeNode) bool {
		if B == nil {
			return true
		}
		if A == nil || A.Val != B.Val {
			return false
		}
		return dfs(A.Left, B.Left) && dfs(A.Right, B.Right)
	}
	return (A != nil && B != nil) && (dfs(A, B) || isSubStructure(A.Left, B) || isSubStructure(A.Right, B))
}
